Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
 
     // TODO: #1001 attribs can be nil here?
     NSDictionary* attribs = [ts->_attributedString attributesAtIndex:range.location effectiveRange:NULL];
-    UIFont* font = [attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)];
+    CTFontRef font = static_cast<CTFontRef>([attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)]);
+
     CGFloat fontSize = kCTFontSystemFontSize;
+    std::vector<wchar_t> fontName(c_defaultFontName, c_defaultFontName + wcslen(c_defaultFontName));
+    DWRITE_FONT_WEIGHT weight = DWRITE_FONT_WEIGHT_REGULAR;
+    DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+    DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
     if (font != nil) {
-        fontSize = [font pointSize];
+        fontSize = CTFontGetSize(font);
+        __InitDWriteFontPropertiesFromName(CTFontCopyName(font, kCTFontFullNameKey), &weight, &stretch, &style);
+        CFStringRef name = _DWriteGetFamilyNameForFontName(CTFontCopyName(font, kCTFontFullNameKey));
+        fontName.resize(CFStringGetLength(name) + 1);
+        CFStringGetCharacters(name, CFRangeMake(0, fontName.size()), reinterpret_cast<UniChar*>(fontName.data()));
     }
 
     ComPtr<IDWriteTextFormat> textFormat;
-    THROW_IF_FAILED(dwriteFactory->CreateTextFormat(c_defaultFontName,
-                                                    NULL,
-                                                    DWRITE_FONT_WEIGHT_REGULAR,
-                                                    DWRITE_FONT_STYLE_NORMAL,
-                                                    DWRITE_FONT_STRETCH_NORMAL,
-                                                    fontSize,
-                                                    c_defaultUserLanguage,
-                                                    &textFormat));
+    THROW_IF_FAILED(
+        dwriteFactory->CreateTextFormat(fontName.data(), NULL, weight, style, stretch, fontSize, c_defaultUserLanguage, &textFormat));
+
+    CTParagraphStyleRef settings =
+        static_cast<CTParagraphStyleRef>([attribs valueForKey:static_cast<NSString*>(kCTParagraphStyleAttributeName)]);
+    if (settings) {
+        CTTextAlignment alignment = kCTNaturalTextAlignment;
+        if (CTParagraphStyleGetValueForSpecifier(settings, kCTParagraphStyleSpecifierAlignment, sizeof(CTTextAlignment), &alignment)) {
+            DWRITE_TEXT_ALIGNMENT dwriteAlignment = DWRITE_TEXT_ALIGNMENT_LEADING;
+            switch (alignment) {
```

Comment:
consider moving this switch statement out into its own helper:
- you won't have to break since you can just early return
- makes it easy to reuse
- breaks the code into smaller chunks